test/gtest/ucs/test_memtype_cache.cc

```diff
@@ -47,38 +64,349 @@ class test_memtype_cache : public ucs::test_with_param<ucs_memory_type_t> {
               << " memtype=" << mem_buffer::mem_type_name(mem_type);
     }
 
+    void test_ptr_found(void *ptr, size_t size,
+                        ucs_memory_type_t expected_type) const {
+        if (expected_type != UCS_MEMORY_TYPE_HOST) {
+            test_lookup_found(ptr, size, expected_type);
+            test_lookup_found(ptr, size / 2, expected_type);
+            test_lookup_found(ptr, 1, expected_type);
+            test_lookup_found(UCS_PTR_BYTE_OFFSET(ptr, size - 1),
+                              1, expected_type);
+            test_lookup_found(ptr, 0, expected_type);
+        }
+    }
+
+    void test_region_found(mem_buffer &b, size_t size) const {
+        test_ptr_found(b.ptr(), size, b.mem_type());
+    }
+
+    void test_region_not_found(mem_buffer &b, size_t size) const {
+        test_ptr_not_found(b.ptr(), size);
+    }
+
+    void test_ptr_not_found(void *ptr, size_t size) const {
+        /* memtype cache is aligned by Page Table defined constant,
+         * so need to step by this value to make something not found */
+        test_lookup_notfound(ptr, size + UCS_PGT_ADDR_ALIGN);
+        test_lookup_notfound(UCS_PTR_BYTE_OFFSET(ptr, size), 1 + UCS_PGT_ADDR_ALIGN);
+    }
+
+    void test_ptr_released(void *ptr, size_t size) const {
+        test_lookup_notfound(ptr, size);
+        test_lookup_notfound(ptr, 1);
+    }
+
+    void simulate_ucm_mem_type_event(void *ptr, size_t size,
+                                     ucs_memory_type_t mem_type,
+                                     ucm_event_type_t event_type) const {
+        ASSERT_TRUE((event_type == UCM_EVENT_MEM_TYPE_ALLOC) ||
+                    (event_type == UCM_EVENT_MEM_TYPE_FREE));
+
+        ucm_event_t event;
+
+        memset(&event, 0, sizeof(event));
+
+        event.mem_type.address  = ptr;
+        event.mem_type.size     = size;
+        event.mem_type.mem_type = mem_type;
+
+        ucm_event_enter();
+        ucm_event_dispatch(event_type, &event);
+        ucm_event_leave();
+    }
+
+    mem_buffer* allocate_mem_buffer(void *ptr, size_t size, ucs_memory_type_t mem_type,
+                                    std::deque<mem_buffer*> *allocated_buffers = NULL,
+                                    bool push_front = false,
+                                    bool test_not_found = true) const {
+        mem_buffer *buf;
+
+        if (ptr != NULL) {
+            buf = new mem_buffer(ptr, size, mem_type);
+            simulate_ucm_mem_type_event(ptr, size, mem_type,
```

Comment:
why do we need to simulate event, wouldn't it be generated already by the allocation itself?

Reply:
actually, no
real allocation doesn't generate a case when different memory types share same memory page
but, now our implementation supports such cases

Reply:
but we should install UCM hooks for cudaMalloc, aren't they sent?

Reply:
> but we should install UCM hooks for cudaMalloc, aren't they sent?

they set inside `ucs_memtype_cache_create()` (i.e. MemoryType Cache constructor)

Reply:
so in this case UCM should fire and event when we are doing cudaFree from mem_buffer release? so we won't need to simulate a UCM event

Reply:
> so in this case UCM should fire and event when we are doing cudaFree from mem_buffer release?

yes

> so we won't need to simulate a UCM event

ok, will remove

Reply:
> so we won't need to simulate a UCM event

done